packages/relay-compiler/transforms/DeclarativeConnectionMutationTransform.js

```diff
@@ -100,55 +115,111 @@ function visitLinkedField(field: LinkedField): LinkedField {
   }
-  const connectionsArg = edgeDirective.args.find(
+  if (edgeDirective != null && nodeDirective != null) {
+    throw createUserError(
+      `Invalid use of @${edgeDirective.name} and @${nodeDirective.name} on field '${transformedField.name}' - these directives cannot be used together.`,
+      [edgeDirective.loc],
+    );
+  }
+  const targetDirective = edgeDirective ?? nodeDirective;
+  const connectionsArg = targetDirective.args.find(
     arg => arg.name === 'connections',
   );
   if (connectionsArg == null) {
     throw createUserError(
-      `Expected the 'connections' argument to be defined on @${edgeDirective.name}.`,
-      [edgeDirective.loc],
+      `Expected the 'connections' argument to be defined on @${targetDirective.name}.`,
+      [targetDirective.loc],
     );
   }
   const schema = this.getContext().getSchema();
-  const fields = schema.getFields(transformedField.type);
-  let cursorFieldID;
-  let nodeFieldID;
-  for (const fieldID of fields) {
-    const fieldName = schema.getFieldName(fieldID);
-    if (fieldName === ConnectionInterface.get().CURSOR) {
-      cursorFieldID = fieldID;
-    } else if (fieldName === ConnectionInterface.get().NODE) {
-      nodeFieldID = fieldID;
+  if (edgeDirective) {
+    const fields = schema.getFields(transformedField.type);
+    let cursorFieldID;
+    let nodeFieldID;
+    for (const fieldID of fields) {
+      const fieldName = schema.getFieldName(fieldID);
+      if (fieldName === ConnectionInterface.get().CURSOR) {
+        cursorFieldID = fieldID;
+      } else if (fieldName === ConnectionInterface.get().NODE) {
+        nodeFieldID = fieldID;
+      }
     }
+
+    // Edge
+    if (cursorFieldID != null && nodeFieldID != null) {
+      const handle: Handle = {
+        name: edgeDirective.name,
+        key: '',
+        dynamicKey: null,
+        filters: null,
+        handleArgs: [connectionsArg],
+      };
+      return {
+        ...transformedField,
+        directives: transformedField.directives.filter(
+          directive => directive !== edgeDirective,
+        ),
+        handles: transformedField.handles
+          ? [...transformedField.handles, handle]
+          : [handle],
+      };
+    }
+    throw createUserError(
+      `Unsupported use of @${edgeDirective.name} on field '${transformedField.name}', expected an edge field (a field with 'cursor' and 'node' selection).`,
+      [targetDirective.loc],
+    );
+  } else {
+    // Node
+    const edgeTypeNameArg = nodeDirective.args.find(
+      arg => arg.name === 'edgeTypeName',
+    );
+    if (!edgeTypeNameArg) {
+      throw createUserError(
+        `Unsupported use of @${nodeDirective.name} on field '${transformedField.name}', 'edgeTypeName' argument must be provided.`,
+        [targetDirective.loc],
+      );
+    }
+    const rawType = schema.getRawType(transformedField.type);
+    if (
+      schema.isObject(rawType) ||
+      schema.isUnion(rawType) ||
+      schema.isInterface(rawType)
```

Comment:
Are these the right constraints?

Reply:
I think `schema.canHaveSelections` is more accurate.

Reply:
Great, thanks! I've fixed that now.